### PR TITLE
ERM parser for wikitables

### PR DESCRIPTION
**Pull request by pdasigi**
This PR adds a wikitables parser that uses ERM to train. Like with nlvr, I split the existing wikitables parser into two parsers: `WikiTablesMmlParser` and `WikiTablesErmParser`, and kept the common functionality in a base class called `WikiTablesParser`. The ERM parser does not have coverage implemented yet. The cost is currently defined only based on denotation errors. I'll open a separate pr for agenda coverage.

### CODE REVIEW

File: allennlp/models/semantic_parsing/wikitables/wikitables_mml_semantic_parser.py [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+    @overrides
+    def forward(self,  # type: ignore
+                question: Dict[str, torch.LongTensor],
+                table: Dict[str, torch.LongTensor],
+                world: List[WikiTablesWorld],
+                actions: List[List[ProductionRuleArray]],
+                example_lisp_string: List[str] = None,
+                target_action_sequences: torch.LongTensor = None) -> Dict[str, torch.Tensor]:
+        # pylint: disable=arguments-differ
+        # pylint: disable=unused-argument
```
Comment (matt-gardner) [inline allennlp/models/semantic_parsing/wikitables/wikitables_mml_semantic_parser.py]: What's the unused argument here?

File: allennlp/models/semantic_parsing/wikitables/wikitables_semantic_parser.py [hunk 2]
```diff
@@ -165,43 +153,16 @@ def forward(self,  # type: ignore
...
-
+        raise NotImplementedError
+
+    def _get_initial_state_and_scores(self,
+                                      question: Dict[str, torch.LongTensor],
+                                      table: Dict[str, torch.LongTensor],
+                                      world: List[WikiTablesWorld],
+                                      actions: List[List[ProductionRuleArray]],
+                                      example_lisp_string: List[str] = None,
+                                      add_world_to_initial_state: bool = False) -> Dict:
```
Comment (matt-gardner) [inline allennlp/models/semantic_parsing/wikitables/wikitables_semantic_parser.py]: Why not just always include the `World`?  It'd make the logic simpler everywhere (e.g., no need for optional inputs to the `State`).

File: allennlp/models/semantic_parsing/wikitables/wikitables_erm_semantic_parser.py [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+        return outputs
+
+    def _get_state_cost(self, state: WikiTablesDecoderState) -> torch.Tensor:
+        if not state.is_finished():
+            raise RuntimeError("_get_state_cost() is not defined for unfinished states!")
+        action_history = state.action_history[0]
+        batch_index = state.batch_indices[0]
+        action_strings = [state.possible_actions[batch_index][i][0] for i in action_history]
+        logical_form = state.world[batch_index].get_logical_form(action_strings)
+        lisp_string = state.example_lisp_string[0]
```
[the remark this answers is not in the document]
Comment (pdasigi) [reply allennlp/models/semantic_parsing/wikitables/wikitables_erm_semantic_parser.py]: Done.

File: allennlp/models/semantic_parsing/wikitables/wikitables_decoder_state.py [hunk 4]
```diff
@@ -51,6 +52,13 @@ class WikiTablesDecoderState(DecoderState['WikiTablesDecoderState']):
         A mapping from flattened entity indices (same as the `values` in the
         ``actions_to_entities`` dictionary) to entity type indices.  This represents what type each
         entity has, which we will use for getting type embeddings in certain circumstances.
+    world : ``List[WikiTablesWorld]``, optional (default=None)
+        The worlds corresponding to grouped elements. We store them here because they're required
```
[the remark this answers is not in the document]
Comment (pdasigi) [reply allennlp/models/semantic_parsing/wikitables/wikitables_decoder_state.py]: Fixed, thanks!

File: allennlp/models/semantic_parsing/wikitables/wikitables_decoder_step.py [hunk 5]
```diff
@@ -503,11 +505,9 @@ def _get_entity_action_logits(self,
         # Shape: (group_size, num_actions)
         action_tensor = Variable(state.score[0].data.new(padded_actions).long())
         type_tensor = Variable(state.score[0].data.new(padded_types).long())
-
```
[the remark this answers is not in the document]
Comment (pdasigi) [reply allennlp/models/semantic_parsing/wikitables/wikitables_decoder_step.py]: Done.

File: allennlp/models/semantic_parsing/wikitables/wikitables_erm_semantic_parser.py [hunk 6]
```diff
@@ -0,0 +1,322 @@
...
+
+    @overrides
+    def forward(self,  # type: ignore
+                question: Dict[str, torch.LongTensor],
+                table: Dict[str, torch.LongTensor],
+                world: List[WikiTablesWorld],
+                actions: List[List[ProductionRuleArray]],
+                example_lisp_string: List[str]) -> Dict[str, torch.Tensor]:
+        # pylint: disable=arguments-differ
+        # pylint: disable=unused-argument
```
[the remark this answers is not in the document]
Comment (pdasigi) [reply allennlp/models/semantic_parsing/wikitables/wikitables_erm_semantic_parser.py]: Nothing. Removed it.

File: allennlp/models/semantic_parsing/wikitables/wikitables_erm_semantic_parser.py [hunk 7]
```diff
@@ -0,0 +1,322 @@
...
+        actions : ``List[List[ProductionRuleArray]]``
+            A list of all possible actions for each ``World`` in the batch, indexed into a
+            ``ProductionRuleArray`` using a ``ProductionRuleField``.  We will embed all of these
+            and use the embeddings to determine which action to take at each timestep in the
+            decoder.
+        example_lisp_string : ``List[str]``
+            The example (lisp-formatted) string corresponding to the given input.  This comes
+            directly from the ``.examples`` file provided with the dataset.  We pass this to SEMPRE
+            when evaluating denotation accuracy; it is otherwise unused.
+        target_action_sequences : torch.Tensor, optional (default=None)
```
[the remark this answers is not in the document]
Comment (pdasigi) [reply allennlp/models/semantic_parsing/wikitables/wikitables_erm_semantic_parser.py]: Done.

File: allennlp/models/semantic_parsing/wikitables/wikitables_erm_semantic_parser.py [hunk 8]
```diff
@@ -0,0 +1,322 @@
...
+                                                          example_lisp_string=example_lisp_string,
+                                                          add_world_to_initial_state=True)
+        initial_state = initial_info["initial_state"]
+        # TODO(pradeep): Keep track of debug info. It's not straightforward currently because the
+        # ERM's decode does not return the best states.
+        outputs = self._decoder_trainer.decode(initial_state,
+                                               self._decoder_step,
+                                               self._get_state_cost)
+        if not self.training:
+            linking_scores = initial_info["linking_scores"]
```
[the remark this answers is not in the document]
Comment (pdasigi) [reply allennlp/models/semantic_parsing/wikitables/wikitables_erm_semantic_parser.py]: The next PR adds agenda coverage computation which makes this block different between the two classes. But I agree that some of these lines can be shared. I added a TODO for this.

File: allennlp/models/semantic_parsing/wikitables/wikitables_mml_semantic_parser.py (hunk 1 above)
Comment (pdasigi) [reply allennlp/models/semantic_parsing/wikitables/wikitables_mml_semantic_parser.py]: Nothing. Removed it.

File: allennlp/models/semantic_parsing/wikitables/wikitables_semantic_parser.py (hunk 2 above)
Comment (pdasigi) [reply allennlp/models/semantic_parsing/wikitables/wikitables_semantic_parser.py]: I thought not including the world could make the code more efficient. The next PR adds checklist stuff which are all optional fields anyway. I'll get back to this when we clean the code later.

**Pull request merged**